Index: hostapd-2016-06-15/wpa_supplicant/events.c
===================================================================
--- hostapd-2016-06-15.orig/wpa_supplicant/events.c	2018-06-14 09:35:03.338441260 +0800
+++ hostapd-2016-06-15/wpa_supplicant/events.c	2018-06-14 09:37:13.718444226 +0800
@@ -983,7 +983,7 @@
 				"   skip - BSSID not in whitelist");
 			continue;
 		}
-
+#if 0
 		if (!wpa_supplicant_ssid_bss_match(wpa_s, ssid, bss))
 			continue;
 
@@ -995,7 +995,7 @@
 				"not allowed");
 			continue;
 		}
-
+#endif
 		if (wpa && !wpa_key_mgmt_wpa(ssid->key_mgmt) &&
 		    has_wep_key(ssid)) {
 			wpa_dbg(wpa_s, MSG_DEBUG, "   skip - ignore WPA/WPA2 AP for WEP network block");
@@ -1007,13 +1007,13 @@
 				"not allowed");
 			continue;
 		}
-
+#if 0
 		if (!wpa_supplicant_match_privacy(bss, ssid)) {
 			wpa_dbg(wpa_s, MSG_DEBUG, "   skip - privacy "
 				"mismatch");
 			continue;
 		}
-
+#endif
 		if (!bss_is_ess(bss) && !bss_is_pbss(bss)) {
 			wpa_dbg(wpa_s, MSG_DEBUG, "   skip - neither ESS nor PBSS network");
 			continue;
