Index: hostapd-2016-06-15/wpa_supplicant/ctrl_iface.c
===================================================================
--- hostapd-2016-06-15.orig/wpa_supplicant/ctrl_iface.c	2018-10-29 16:14:21.644743286 +0800
+++ hostapd-2016-06-15/wpa_supplicant/ctrl_iface.c	2018-10-29 16:15:57.412745038 +0800
@@ -3193,7 +3193,7 @@
 		return -1;
 	}
 
-	value = wpa_config_get_no_key(ssid, name);
+	value = wpa_config_get(ssid, name);
 	if (value == NULL) {
 		wpa_printf(MSG_EXCESSIVE, "CTRL_IFACE: Failed to get network "
 			   "variable '%s'", name);
