--- a/src/mod/endpoints/mod_verto/Makefile.am
+++ b/src/mod/endpoints/mod_verto/Makefile.am
@@ -12,8 +12,8 @@ if HAVE_PERL
 #perldir = $(PERL_SITEDIR)
 noinst_LTLIBRARIES = MCAST.la
 MCAST_la_SOURCES   = mcast/mcast_wrap.cpp mcast/perlxsi.c mcast/mcast.c mcast/mcast_cpp.cpp
-MCAST_la_CFLAGS    = $(CC_CFLAGS) $(CFLAGS) $(SWITCH_AM_CFLAGS) $(PERL_CFLAGS)
-MCAST_la_CXXFLAGS  = $(SWITCH_AM_CXXFLAGS) $(CXXFLAGS) -w $(PERL_INC)
+MCAST_la_CFLAGS    = $(CC_CFLAGS) $(SWITCH_AM_CFLAGS) $(PERL_CFLAGS)
+MCAST_la_CXXFLAGS  = $(SWITCH_AM_CXXFLAGS) -w $(PERL_INC)
 MCAST_la_CPPFLAGS  = -I$(switch_srcdir)/src/mod/endpoints/mod_verto/mcast
 MCAST_la_LDFLAGS   = -avoid-version -module -no-undefined -shared $(PERL_LDFLAGS)
 
