Index: pygobject-2.16.0/glib/pyglib.c
===================================================================
--- pygobject-2.16.0.orig/glib/pyglib.c
+++ pygobject-2.16.0/glib/pyglib.c
@@ -359,7 +359,7 @@ pyglib_gerror_exception_check(GError **e
     }
 
     g_set_error(error, g_quark_from_string(_PyUnicode_AsString(py_domain)),
-                _PyLong_AsLong(py_code), _PyUnicode_AsString(py_message));
+                _PyLong_AsLong(py_code), "%s", _PyUnicode_AsString(py_message));
 
     Py_DECREF(py_message);
     Py_DECREF(py_code);
@@ -368,7 +368,7 @@ pyglib_gerror_exception_check(GError **e
 
 bad_gerror:
     Py_DECREF(value);
-    g_set_error(error, g_quark_from_static_string("pyglib"), 0, bad_gerror_message);
+    g_set_error(error, g_quark_from_static_string("pyglib"), 0, "%s", bad_gerror_message);
     PyErr_SetString(PyExc_ValueError, bad_gerror_message);
     PyErr_Print();
     return -2;
Index: pygobject-2.16.0/glib/pygspawn.c
===================================================================
--- pygobject-2.16.0.orig/glib/pygspawn.c
+++ pygobject-2.16.0/glib/pygspawn.c
@@ -22,7 +22,7 @@
  */
 
 #include <Python.h>
-#include <glib/gspawn.h>
+#include <glib.h>
 
 #include "pyglib.h"
 #include "pyglib-private.h"
