--- a/src/mod/languages/mod_python/Makefile.am
+++ b/src/mod/languages/mod_python/Makefile.am
@@ -20,7 +20,7 @@ swigclean: clean
 	rm -f mod_python_wrap.* freeswitch.py
 
 mod_python_wrap.cpp: mod_python_extra.c hack.diff 
-	swig2.0 -python -c++ -I../../../../src/include -oh mod_python_wrap.h -o mod_python_wrap.cpp mod_python.i
+	${STAGING_DIR}/host/bin/swig -python -c++ -I../../../../src/include -oh mod_python_wrap.h -o mod_python_wrap.cpp mod_python.i
 	echo "#include \"mod_python_extra.c\"" >> mod_python_wrap.cpp
 	patch -s -p0 -i hack.diff
 
