--- a/drivers/dahdi/Kbuild
+++ b/drivers/dahdi/Kbuild
@@ -61,9 +61,8 @@ obj-m += $(DAHDI_MODULES_EXTRA)
 # If you want to build OSLEC, include the code in the standard location:
 # drivers/staging/echo . The DAHDI OSLEC echo canceller will be built as
 # well:
-ifneq (,$(wildcard $(src)/../staging/echo/echo.c))
+ifneq (,$(wildcard $(src)/../staging/echo/oslec.h))
 obj-m += dahdi_echocan_oslec.o
-obj-m += ../staging/echo/echo.o
 endif
 
 CFLAGS_MODULE += -I$(DAHDI_INCLUDE) -I$(src)
