Index: libpri-1.4.15/pritest.c
===================================================================
--- libpri-1.4.15.orig/pritest.c
+++ libpri-1.4.15/pritest.c
@@ -41,7 +41,7 @@
 #include <sys/ioctl.h>
 #include <stdlib.h>
 #include <unistd.h>
-#include <sys/signal.h>
+#include <signal.h>
 #include <sys/select.h>
 #include <sys/wait.h>
 #include <sys/resource.h>
Index: libpri-1.4.15/testprilib.c
===================================================================
--- libpri-1.4.15.orig/testprilib.c
+++ libpri-1.4.15/testprilib.c
@@ -41,7 +41,7 @@
 #include <sys/ioctl.h>
 #include <stdlib.h>
 #include <unistd.h>
-#include <sys/signal.h>
+#include <signal.h>
 #include <sys/select.h>
 #include <sys/wait.h>
 #include <sys/resource.h>
