--- a/autoconf/extra.m4
+++ b/autoconf/extra.m4
@@ -472,17 +472,6 @@ AC_DEFUN([CS_ENABLE_OPTIMIZATION], [
 		])
 	 	CPPFLAGS_saved="${CPPFLAGS_saved} -D_FORTIFY_SOURCE=2"
 		GDB_FLAGS=""
-	], [
-	 	CFLAGS_saved="`echo ${CFLAGS_saved} |sed -e 's/\-O[0-9]\ \?//g' -e 's/[^|\ ]\-g[$|\ ]//g'`"
-		optimize_flag="-O0"
-		case "${CC}" in
-			*gcc*)
-				AX_CHECK_COMPILE_FLAG(-Og, [
-					optimize_flag="-Og"
-				])
-			;;
-		esac
-		CFLAGS_saved="${CFLAGS_saved} ${optimize_flag} "
 	])
 
 	AS_IF([test "X${enable_debug}" == "Xyes"], [
