--- a/squashfs-tools/mksquashfs.c	2018-08-07 14:43:09.861017163 +0800
+++ b/squashfs-tools/mksquashfs.c	2018-08-07 14:44:48.857019415 +0800
@@ -824,13 +824,13 @@
 }
 
 
-inline unsigned int get_inode_no(struct inode_info *inode)
+ unsigned int get_inode_no(struct inode_info *inode)
 {
 	return inode->inode_number;
 }
 
 
-inline unsigned int get_parent_no(struct dir_info *dir)
+ unsigned int get_parent_no(struct dir_info *dir)
 {
 	return dir->depth ? get_inode_no(dir->dir_ent->inode) : inode_no;
 }
@@ -2053,7 +2053,7 @@
 }
 
 
-inline int is_fragment(struct inode_info *inode)
+ int is_fragment(struct inode_info *inode)
 {
 	int file_size = inode->buf.st_size;
 
@@ -3016,20 +3016,20 @@
 }
 
 
-inline struct inode_info *lookup_inode(struct stat *buf)
+ struct inode_info *lookup_inode(struct stat *buf)
 {
 	return lookup_inode2(buf, 0, 0);
 }
 
 
-inline void alloc_inode_no(struct inode_info *inode, unsigned int use_this)
+ void alloc_inode_no(struct inode_info *inode, unsigned int use_this)
 {
 	if (inode->inode_number == 0)
 		inode->inode_number = use_this ? : inode_no ++;
 }
 
 
-inline struct dir_ent *create_dir_entry(char *name, char *source_name,
+ struct dir_ent *create_dir_entry(char *name, char *source_name,
 	char *nonstandard_pathname, struct dir_info *dir)
 {
 	struct dir_ent *dir_ent = malloc(sizeof(struct dir_ent));
@@ -3046,7 +3046,7 @@
 }
 
 
-inline void add_dir_entry(struct dir_ent *dir_ent, struct dir_info *sub_dir,
+ void add_dir_entry(struct dir_ent *dir_ent, struct dir_info *sub_dir,
 	struct inode_info *inode_info)
 {
 	struct dir_info *dir = dir_ent->our_dir;
@@ -3062,7 +3062,7 @@
 }
 
 
-inline void add_dir_entry2(char *name, char *source_name,
+ void add_dir_entry2(char *name, char *source_name,
 	char *nonstandard_pathname, struct dir_info *sub_dir,
 	struct inode_info *inode_info, struct dir_info *dir)
 {
@@ -3074,7 +3074,7 @@
 }
 
 
-inline void free_dir_entry(struct dir_ent *dir_ent)
+ void free_dir_entry(struct dir_ent *dir_ent)
 {
 	if(dir_ent->name)
 		free(dir_ent->name);
@@ -3086,7 +3086,7 @@
 }
 
 
-inline void add_excluded(struct dir_info *dir)
+ void add_excluded(struct dir_info *dir)
 {
 	dir->excluded ++;
 }
